### PR TITLE
Matthew/bce loss

### PR DESCRIPTION
**Pull request by growlix**
Short-term solution for adding binary cross entropy loss to `ComposerClassifier` (as [per](https://mosaicml.slack.com/archives/C036RMYG1EH/p1650059048931689)). Roughly, the implementation is to add a `loss` (str) param to a model and its associated hparams (e.g. `resnet_hparams` and `ComposerResNet`), which then gets passed through to `ComposerClassifier` and set as an attribute. Loss functions are defined in `loss.loss.py` and added to a `loss_registry` dict, which is how they're checked/accessed.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -110,6 +110,7 @@ venv/
 ENV/
 env.bak/
 venv.bak/
+list/
```
Comment (ravi-mosaicml) [inline .gitignore:113]: What is the `list/` folder?

Comment (growlix) [reply .gitignore]: Was created as part of a virtual environment. Not sure how that got there, but it's been removed.

**Review by Landanjs — COMMENTED** [review]
This seems like a good setup given how composer is stuctured right now, but it still makes me comfortable for two reasons:
1. Copying over the bce function signature: this wouldn't be necessary if the loss calculation and the reduction mechanisms were separated... Design discussion for another day though?
2. BCE is unlikely to work well unless the bias is set to `-log(num_classes)`. I feel like many users will miss this detail, but I don't know if there will be a good way to enforce this automatically (in the model definition, not parent class).

Is there a better way to implement the necessary initialization? There are cases where other linear layers are in the network (ViT) or when the final layer is a 1x1 conv instead (DeepLabv3). Should we consider the initialization necessary to use BCE? It seems so

Overall, this looks functional for many cases, but there are known cases (ViT and Deeplabv3) where different initializations are needed. I would want to see @ravi-mosaicml opinion

File: composer/loss/loss.py [hunk 2]
```diff
@@ -9,9 +9,68 @@
...
 from torch.nn import functional as F
 
-from composer.loss.utils import infer_target_type
+from composer.loss.utils import ensure_targets_one_hot, infer_target_type
 
-__all__ = ["soft_cross_entropy"]
+__all__ = ["bce", "loss_registry", "soft_cross_entropy"]
+
+
+def bce(
```
[the remark this answers is not in the document]
Comment (growlix) [reply composer/loss/loss.py:19]: Renamed to `binary_cross_entropy_with_logits`!

File: composer/loss/loss.py [hunk 3]
```diff
@@ -9,9 +9,68 @@
...
+            specifying either of those two args will override ``reduction``. Default:
+            ``'sum'``
+        pos_weight (Tensor, optional): a weight of positive examples.
+                Must be a vector with length equal to the number of classes.
+        scale_by_batch_size (bool, optional): Whether to scale the loss by the batch size
+            (i.e. input.shape[0]). Default: ``True``.
+    """
+    target = ensure_targets_one_hot(input, target)
+    l = F.binary_cross_entropy_with_logits(input, target, weight, size_average, reduce, reduction, pos_weight)
+    if scale_by_batch_size:
```
[the remark this answers is not in the document]
Comment (growlix) [reply composer/loss/loss.py:73]: Are you suggesting something instead like:

```
bce = F.binary_cross_entropy_with_logits(input, target, weight, size_average, reduce, reduction, pos_weight)
if reduction == 'sum':
    bce /= torch.tensor(input.shape[0])
```

File: composer/models/tasks/classification.py [hunk 4]
```diff
@@ -45,12 +48,16 @@ class ComposerClassifier(ComposerModel):
...
+    def __init__(self, module: torch.nn.Module, loss: str = "soft_cross_entropy") -> None:
         super().__init__()
         self.train_acc = Accuracy()
         self.val_acc = Accuracy()
         self.val_loss = CrossEntropy()
         self.module = module
+        if loss not in loss_registry.keys():
+            raise ValueError(f"Unrecognized loss function: {loss}. Please ensure the "
+                             "specified loss function is present in composer.loss.loss.py")
+        self._loss_fxn = loss_registry[loss]
```
[the remark this answers is not in the document]
Comment (growlix) [reply composer/models/tasks/classification.py:60]: Warning thrown!

File: composer/models/tasks/classification.py [hunk 5]
```diff
@@ -45,23 +51,34 @@ class ComposerClassifier(ComposerModel):
...
         if hasattr(self.module, "num_classes"):
             self.num_classes = getattr(self.module, "num_classes")
 
+        if loss_name == 'binary_cross_entropy_with_logits':
+            log.warning("UserWarning: Using `binary_cross_entropy_loss_with_logits` "
+                        "without using `initializers.linear_log_constant_bias` can degrade "
+                        "performance. "
+                        "Please ensure you are using `initializers. "
+                        "linear_log_constant_bias`.")
+
```
Comment (Landanjs) [inline composer/models/tasks/classification.py:74]: I think Ravi likes to use textwrap.dedent(""" """) for these long strings, but probably not necessary?

**Review by Landanjs — APPROVED** [review]
LGTM
